### PR TITLE
BigQuery Driver Implementation

### LINKED ISSUE #33: Add support for BigQuery
## Problem
I want to use monosi with my BigQuery data store

## Solution
Monosi needs to interpret BigQuery configuration details and run on BigQuery accordingly

## Requirements
1. BigQuery instance
2. Refactor to support multiple formats for configurations
3. Add BigQuery driver & dialect

### PR DESCRIPTION
**Pull request by iporollo**
## Pull request type

Please check the type of change your PR introduces:

- [x] Feature

## Issue

Closes #33 

## Description of change & new behavior

Adds support for BigQuery driver & dialect.

## Screenshots

_For frontend updates, please include a screenshot._

## Technical Spec/Implementation Notes

The initial implementation uses an authentication method through the `service.json` file. 

image

### CODE REVIEW

File: src/ingestion/sources/bigquery.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            SELECT 
+                CURRENT_TIMESTAMP as `WINDOW_START`, 
+                CURRENT_TIMESTAMP as `WINDOW_END`, 
+                COUNT(*) as `ROW_COUNT`, 
+                '{table}' as `TABLE_NAME`,
+                '{database}' as `DATABASE_NAME`,
+                '{schema}' as `SCHEMA_NAME`,
+
+                {select_sql}
+            FROM {schema}.{table};
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:20]: Does this need database identifier?

Is there something else about this query that is different from the base?

File: src/ingestion/sources/bigquery.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                COUNT(*) as `ROW_COUNT`, 
+                {table} as `TABLE_NAME`,
+                {database} as `DATABASE_NAME`,
+                {schema} as `SCHEMA_NAME`,
+                {timestamp_field} as `TIMESTAMP_FIELD`,
+
+                {select_sql}
+            FROM {schema}.{table} as c
+            WHERE 
+                DATE_TRUNC({timestamp_field}, HOUR) >= TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL {minutes_ago} MINUTE)
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:42]: CURRENT_TIMESTAMP() here vs. CURRENT_TIMESTAMP above

Did we check that backfill query executes correctly? If not, we should just not include it for now

File: src/ingestion/sources/bigquery.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            credentials_base64=configuration.get('credentials_base64')
+        )
+
+        connection_string = '{connection_string_prefix}://{connection_string_base}{connection_string_params}'.format(
+            connection_string_prefix=connection_string_prefix,
+            connection_string_base=connection_string_base,
+            connection_string_params=connection_string_params
+        )
+
+        return make_url(connection_string)
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:90]: what is make_url doing here?

Also why not just use https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.engine.URL.create

If you're already using a sqlalchemy helper and the configuration of the string looks really hard to follow, it might be a bit more elegant

File: src/ingestion/sources/bigquery.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+
+    def database(self):
+        return json.loads(self.configuration).get("project")
+
+    def schema(self):
+        return json.loads(self.configuration).get('dataset')
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:97]: This feels dangerous - it actually works this way?

File: src/ingestion/sources/bigquery.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    @classmethod
+    def table_metrics_query(cls, monitor, discovery_data, minutes_ago):
+        builder = BigQueryMetricsQueryBuilder(cls, monitor, discovery_data, minutes_ago)
+        query = builder.compile()
+        return query
+
+
+    @classmethod
+    def schema_columns_query(cls, database_name, schema_name):
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:160]: Why do we need to override this?

File: src/ingestion/sources/bigquery.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    @classmethod
+    def _freshness(cls):
+        return "TIMESTAMP_DIFF(MAX({}), CURRENT_TIMESTAMP, MINUTE)"
+
+    @classmethod
+    def schema_tables_query(cls, database_name, schema_name):
+        raise NotImplementedError
+
+    @classmethod
+    def table_metrics_query(cls, monitor, discovery_data, minutes_ago):
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:153]: Why do we need to override this?

File: src/ingestion/sources/bigquery.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            AND LOWER(c.table_catalog) = LOWER('{database_name}')
+        """.format(database_name=database_name, schema_name=schema_name)
+
+    @classmethod
+    def query_access_logs_query(cls):
+        raise NotImplementedError
+        
+    @classmethod
+    def query_copy_logs_query(cls):
+        raise NotImplementedError
```
Comment (unkrich) [inline src/ingestion/sources/bigquery.py:186]: theoretically should have notimplemented on base and let it fall through and move implementations to corresponding ones that work

File: src/ui/src/components/forms/DatasourceForm/index.tsx [hunk 8]
```diff
@@ -38,49 +40,87 @@ const DatasourceForm = () => {
...
+        user: user,
+        password: password,
+        host: host,
+        port: parseInt(port),
+        database: database,
+        schema: schema
+      };
+    } else if (datasourceType === DataSourceTypes.BIGQUERY) {
+      // TODO: Implement better validation
+      if(!datasourceName || !project || !dataset || !file) return;
```
Comment (unkrich) [inline src/ui/src/components/forms/DatasourceForm/index.tsx:75]: why add validation here but not on any others?

File: src/ui/src/components/forms/DatasourceForm/index.tsx [hunk 9]
```diff
@@ -301,6 +353,46 @@ const DatasourceForm = () => {
           </EuiFormRow>
         </div>
       )}
+      {datasourceType === DataSourceTypes.BIGQUERY && (
+        <div>
+          <EuiPageHeader
+            iconType="https://cdn.worldvectorlogo.com/logos/google-bigquery-logo-1.svg"
```
Comment (unkrich) [inline src/ui/src/components/forms/DatasourceForm/index.tsx:359]: You added this file, should link to that

File: src/ui/src/images/index.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+import PagerDutyLogo from './PagerDutyLogo';
```
Comment (unkrich) [inline src/ui/src/images/index.ts:1]: Is there a reason to go from tsx to ts?

File: src/ingestion/sources/bigquery.py (hunk 4 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:97]: Yes - BigQuery has a concept of projects and datasets. Under the hood, those represent `catalog` and `schema` respectively, which can be associated with database and schema in our systems.

File: src/ui/src/components/forms/DatasourceForm/index.tsx (hunk 8 above)
Comment (iporollo) [reply src/ui/src/components/forms/DatasourceForm/index.tsx:75]: because adding validation to other seemed out of scope of the PR, hence added it only to BigQuery.

File: src/ui/src/images/index.ts (hunk 10 above)
Comment (iporollo) [reply src/ui/src/images/index.ts:1]: because the index file is handling imports and will never contain `tsx` syntax, hence not necessary

File: src/ingestion/sources/bigquery.py (hunk 1 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:20]: Not necessary but will add for consistency. Yes, the way BigQuery takes string literal column names is with back ticks rather than quotes, thus those change as well

File: src/ingestion/sources/bigquery.py (hunk 2 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:42]: Yes the query was checked and executes successfully. `CURRENT_TIMESTAMP` and `CURRENT_TIMESTAMP()` both pull the timestamp in BQ, but we will use `CURRENT_TIMESTAMP()` per the docs recommendations - https://cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions

File: src/ingestion/sources/bigquery.py (hunk 6 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:153]: because we need to use the `BigQueryMetricsQueryBuilder` class for the builder

File: src/ingestion/sources/bigquery.py (hunk 5 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:160]: Because the way BigQuery takes string literal column names is with back ticks rather than quotes, thus those need to be overwritten

File: src/ingestion/sources/bigquery.py (hunk 7 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:186]: sounds good, removing

File: src/ui/src/components/forms/DatasourceForm/index.tsx (hunk 9 above)
Comment (iporollo) [reply src/ui/src/components/forms/DatasourceForm/index.tsx:359]: good catch, replaced it in one spot, not the other

File: src/ingestion/sources/bigquery.py (hunk 3 above)
Comment (iporollo) [reply src/ingestion/sources/bigquery.py:90]: Using `make_url()` for sanity to create a proper URL for the SQLAlchemy engine. 

I don't believe we can use `URL.create` because the parameters for creating a BQ connection string are not standard

Comment (unkrich) [reply src/ingestion/sources/bigquery.py:90]: Disagree, but not going to block

File: src/ui/src/components/forms/DatasourceForm/index.tsx (hunk 8 above)
Comment (unkrich) [reply src/ui/src/components/forms/DatasourceForm/index.tsx:75]: Yeah... having it or not having it with this implementation will still produce confusing behaviors either way so I suppose it really doesn't matter given that it's still not done in a way that handles user input correctly

File: src/ingestion/sources/bigquery.py (hunk 2 above)
Comment (unkrich) [reply src/ingestion/sources/bigquery.py:42]: should include database here as well

**Review by unkrich — APPROVED** [review]
